lib/galaxy/config/__init__.py

```diff
@@ -180,57 +173,142 @@ def _parse_config_file_options(self, defaults, listify_defaults, config_kwargs):
                             break
                     else:
                         paths = listify(value)
                         break
                 else:
                     paths = listify(values[-1])
                 setattr(self, var + '_set', False)
-            setattr(self, var, [resolve_path(x, self.root) for x in paths])
+            setattr(self, var, [os.path.join(self.root, x) for x in paths])
 
 
 class GalaxyAppConfiguration(BaseAppConfiguration):
     deprecated_options = ('database_file', 'track_jobs_in_database')
     default_config_file_name = 'galaxy.yml'
+    # {key: config option, value: deprecated directory name}
+    # If value == first dir in a user path that resolves to key, it will be stripped from the path
+    deprecated_dirs = {'config_dir': 'config', 'data_dir': 'database'}
 
     def __init__(self, **kwargs):
         self._load_schema()  # Load schema from schema definition file
         self._load_config_from_schema()  # Load default propery values from schema
+        self._validate_schema_paths()  # check that paths can be resolved
         self._update_raw_config_from_kwargs(kwargs)  # Overwrite default values passed as kwargs
         self._create_attributes_from_raw_config()  # Create attributes for LOADED properties
+        self._resolve_paths(kwargs)  # Overwrite attributes (not _raw_config) w/resolved paths
         self._process_config(kwargs)  # Finish processing configuration
 
     def _load_schema(self):
         self.schema = AppSchema(GALAXY_CONFIG_SCHEMA_PATH, GALAXY_APP_NAME)
         self.appschema = self.schema.app_schema
 
     def _load_config_from_schema(self):
-        self._raw_config = {}  # keeps track of values provided to the app
-        self.reloadable_options = set()
+        self._raw_config = {}  # keeps track of startup values (kwargs or schema default)
+        self.reloadable_options = set()  # config options we can reload at runtime
+        self._paths_to_resolve = {}  # {config option: referenced config option}
         for key, data in self.appschema.items():
             self._raw_config[key] = data.get('default')
             if data.get('reloadable'):
                 self.reloadable_options.add(key)
+            if data.get('path_resolves_to'):
+                self._paths_to_resolve[key] = data.get('path_resolves_to')
+
+    def _validate_schema_paths(self):
+
+        def check_exists(option, key):
+            if not option:
+                message = "Invalid schema: property '{}' listed as path resolution target " \
+                    "for '{}' does not exist".format(resolves_to, key)
+                raise_error(message)
+
+        def check_type_is_str(option, key):
+            if option.get('type') != 'str':
+                message = "Invalid schema: property '{}' should have type 'str'".format(key)
+                raise_error(message)
+
+        def check_is_dag():
+            def walk(key):
+                if key:
```

Comment:
Is that check needed?

Reply:
Yes. If key is None, we've reached the base case: None will be returned and recursion terminates.

Reply:
Ah, OK, then I'd still remove it from here and do the check before recursing (presently line 238).

Reply:
Like this?
```python
 def walk(key):
     if key in visited:                                                            
         raise_error('Invalid schema: cycle detected')                             
     visited.add(key)
     next_key = self.appschema[key].get('path_resolves_to')                        
     if next_key:
         walk(next_key)  
```
I don't mind. To me, testing for the base case at the top of the function makes it more intuitive (that's how my reasoning would work), but I don't mind doing it this way.

Reply:
Yes, see the code suggestions below!